File: src/main/java/com/campspot/resources/PunchResource.java
package com.campspot.resources;

import com.campspot.api.Punch;
import com.campspot.lib.PunchLib;
import com.codahale.metrics.annotation.Timed;
import io.dropwizard.jersey.params.DateTimeParam;
import org.hibernate.validator.valuehandling.UnwrapValidatedValue;
import org.skife.jdbi.v2.TransactionIsolationLevel;
import org.skife.jdbi.v2.sqlobject.Transaction;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.QueryParam;
import javax.ws.rs.core.MediaType;
import java.util.List;

@Path("/punches")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class PunchResource {
  private final PunchLib punchLib;

  public PunchResource(PunchLib punchLib) {
    this.punchLib = punchLib;
  }

  @POST
  @Timed
  @Valid
  @Transaction(TransactionIsolationLevel.READ_COMMITTED)
  public Punch create(@Valid @NotNull Punch punch) {
    return punchLib.create(punch);
  }

  @PUT
  @Timed
  @Valid
  @Transaction(TransactionIsolationLevel.READ_COMMITTED)
  public Punch update(@Valid @NotNull Punch punch) {
    return punchLib.update(punch);
  }

  @GET
  @Timed
  @Valid
  @Transaction(TransactionIsolationLevel.READ_COMMITTED)
  public List<Punch> list(
    @NotNull @UnwrapValidatedValue @QueryParam("start") DateTimeParam start,
    @NotNull @UnwrapValidatedValue @QueryParam("end") DateTimeParam end
  ) {
    return punchLib.listForDates(start.get(), end.get());
  }
}

File: src/main/kotlin/com/campspot/resources/PunchResource.kt
package com.campspot.resources

import com.campspot.TimeTrackingApplication.Companion.MASTER
import com.campspot.api.Punch
import com.campspot.jdbi3.InTransaction
import com.campspot.lib.PunchLib
import com.codahale.metrics.annotation.Timed
import io.dropwizard.jersey.jsr310.ZonedDateTimeParam
import org.hibernate.validator.valuehandling.UnwrapValidatedValue
import javax.validation.Valid
import javax.validation.constraints.NotNull
import javax.ws.rs.*
import javax.ws.rs.core.MediaType

@Path("/punches")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
class PunchResource(private val punchLib: PunchLib) {

  @POST
  @Timed
  @Valid
  @InTransaction(name = MASTER)
  fun create(@Valid @NotNull punch: Punch): Punch {
    return punchLib.create(punch)
  }

  @PUT
  @Timed
  @Valid
  @InTransaction(name = MASTER)
  fun update(@Valid @NotNull punch: Punch): Punch {
    return punchLib.update(punch)
  }

  @GET
  @Timed
  @Valid
  @InTransaction(name = MASTER)
  fun list(
    @NotNull @UnwrapValidatedValue @QueryParam("start") start: ZonedDateTimeParam,
    @NotNull @UnwrapValidatedValue @QueryParam("end") end: ZonedDateTimeParam
  ): List<Punch> {
    return punchLib.listForDates(start.get(), end.get())
  }
}
